### DISCUSSION TITLE
ROHD swizzle output variable must be equal to the swizzle result width

### DISCUSSION THREAD

**Discussion by quekyj** [intel/rohd#308] [Q&A] [answered]
In ROHD, the `width` of the `swizzle` must be equal to the `width` of the output. The code below fail with error message `Uncaught Error: Failed to put value on signal (f): Updated value width mismatch. The width of 10'bzzzz00111z should be 11`.

Will there be support that allows `0` to be added Infront if the width of the output is larger than the swizzle? 

```dart
import 'package:rohd/rohd.dart';

void slicing(Logic a, Logic b, Logic c, Logic d, Logic e, Logic f) {
  d <= b[7];
  e <= [d, c, a].swizzle();
  f <= [d, c, a].rswizzle();
}

void main() async {
  // Declare Logic
  final a = Logic(name: 'a', width: 4);
  final b = Logic(name: 'b', width: 8);
  final c = Const(7, width: 5);
  final d = Logic(name: 'd');
  final e = Logic(name: 'e', width: d.width + c.width + a.width);
  
  // PURPOSELY ADD +1 here.
  final f = Logic(name: 'f', width: d.width + c.width + a.width + 1);
  
  final rangeSwizzling = RangeSwizzling(a, b, c, d, e, f, slicing);
  await rangeSwizzling.build();
  print(rangeSwizzling.generateSynth());
  
  print('\n');
  a.put(bin('1110'));

  b.put(bin('11000100'));
  print('e: ${rangeSwizzling.e.value.toString(includeWidth: false)}');
  print('f: ${rangeSwizzling.f.value.toString(includeWidth: false)}');
}


class RangeSwizzling extends Module {
  Logic get d => output('d');
  Logic get e => output('e');
  Logic get f => output('f');
  RangeSwizzling(Logic a, Logic b, Logic c, Logic d, Logic e, Logic f,
      void Function(Logic a, Logic b, Logic c, Logic d, Logic e, Logic f) slice)
      : super(name: 'range_swizzling') {
    a = addInput(a.name, a, width: a.width);
    b = addInput(b.name, b, width: b.width);
    c = addInput(c.name, c, width: c.width);
    d = addOutput(d.name, width: d.width);
    e = addOutput(e.name, width: e.width);
    f = addOutput(f.name, width: f.width);
    slice(a, b, c, d, e, f);
  }
}
```

**Comment by chykon** [comment, marked as answer]
There is a method [zeroExtend](https://pub.dev/documentation/rohd/latest/rohd/Logic/zeroExtend.html). Or have I misunderstood something?

```dart
f <= [d, c, a].rswizzle().zeroExtend(f.width);
```

**Reply by quekyj** [reply]
I see. This is exactly what I want. Just a quick question, should `swizzle()` always `zeroExtend` based on its own logic width?

**Reply by chykon** [reply]
My opinion is to use `zeroExtend`/`signExtend` only when explicitly required. ROHD will not allow a circuit with conflicting widths to work.

**Comment by mkorbel1** [comment]
This code actually should fail at connection time rather than `put` time.  I tried running it and got this output:
```
Exception: Bus widths must match. Cannot connect Logic(11): f to Logic(10): _swizzled which have different widths.
```

But the root-cause is the same, and the other answer from @chykon still applies.

**Reply by chykon** [reply]
Hmm, [DartPad](https://dartpad.dev/?id=cf3b17d840ec5005a654e16699bc6661) shows the same error as @quekyj

**Reply by mkorbel1** [reply]
That's weird!  I wonder why

**Reply by mkorbel1** [reply]
Ah, I was running on a side-branch of ROHD which had some different behavior.  The old error should be reinstated since I think it's much more helpful!

**Reply by mkorbel1** [reply]
I filed a bug here: https://github.com/intel/rohd/issues/311
